src/Storages/MergeTree/MergeTreeReadTask.cpp

```diff
@@ -173,13 +156,21 @@ MergeTreeReadTask::Readers MergeTreeReadTask::createReaders(
     for (const auto & pre_columns_per_step : read_info->task_columns.pre_columns)
     {
         if (const auto * index_read_task = getIndexReadTaskForReadStep(read_info->index_read_tasks, pre_columns_per_step))
+        {
+            /// Do not skip marks for queries with FINAL in the reader,
+            /// because it may affect the result of the merging algorithm.
+            bool can_skip_marks = !index_read_task->is_final;
```

Comment:
This means even in case of the `MergeTree()` engine, we cannot skip the mark in `FINAL` query. Why is it the case with V3?